### PR TITLE
boards: Move -fno-strict-aliasing from Make.defs to Toolchain.defs

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Summary
and migrate MAXOPTIMIZATION into ARCHOPTIMIZATION

## Impact
Code refactor

## Testing
Pass CI

### CODE REVIEW

File: arch/arm/src/arm/Toolchain.defs [hunk 1]
```diff
@@ -51,20 +51,22 @@ endif
 #  CROSSDEV         The GNU toolchain triple (command prefix)
 #  ARCHCPUFLAGS     CPU-specific flags selecting the instruction set
 #                   FPU options, etc.
-#  MAXOPTIMIZATION  The maximum optimization level that results in
+#  ARCHOPTIMIZATION The maximum optimization level that results in
```
Comment (pkarashchenko) [inline arch/arm/src/arm/Toolchain.defs:54]: is it still the `The maximum optimization level`?

File: Documentation/guides/cpp_cmake.rst [hunk 2]
```diff
@@ -88,7 +88,7 @@ Creating the project
 
     include(cmake/stm32f4discovery.cmake)
 
-    set(AC_COMMON_FLAGS "${AC_COMMON_FLAGS} -Wall -Wshadow -Wundef -fno-strict-aliasing -Os")
+    set(AC_COMMON_FLAGS "${AC_COMMON_FLAGS} -Wall -Wshadow -Wundef -Os")
```
Comment (pkarashchenko) [inline Documentation/guides/cpp_cmake.rst:91]: why removed from here?

Comment (xiaoxiang781216) [reply Documentation/guides/cpp_cmake.rst:91]: revert.

File: arch/arm/src/arm/Toolchain.defs (hunk 1 above)
Comment (xiaoxiang781216) [reply arch/arm/src/arm/Toolchain.defs]: Remove "maximum"

File: arch/arm/src/arm/Toolchain.defs [hunk 3]
```diff
@@ -51,20 +51,22 @@ endif
 #  CROSSDEV         The GNU toolchain triple (command prefix)
 #  ARCHCPUFLAGS     CPU-specific flags selecting the instruction set
 #                   FPU options, etc.
-#  MAXOPTIMIZATION  The maximum optimization level that results in
+#  ARCHOPTIMIZATION The optimization level that results in
 #                   reliable code generation.
 #
 
+ARCHOPTIMIZATION += -fno-strict-aliasing
```
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply arch/arm/src/arm/Toolchain.defs]: Add the check back.

File: boards/sim/sim/sim/scripts/Make.defs [hunk 4]
```diff
@@ -41,18 +41,18 @@ ARCHDEFINES += -U__NetBSD__ -U__linux__ -U__sun__ -U__unix__
...
-  ARCHOPTIMIZATIONRUST = -g
+  ARCHOPTIMIZATION += -g
+  ARCHOPTIMIZATIONRUST += -g
 endif
 
 ifeq ($(CONFIG_DEBUG_CUSTOMOPT),y)
-  MAXOPTIMIZATION := $(CONFIG_DEBUG_OPTLEVEL)
-else
-  MAXOPTIMIZATION ?= -O2
+  ARCHOPTIMIZATION :+= $(CONFIG_DEBUG_OPTLEVEL)
```
Comment (pkarashchenko) [inline boards/sim/sim/sim/scripts/Make.defs:49]:
```suggestion
  ARCHOPTIMIZATION += $(CONFIG_DEBUG_OPTLEVEL)
```

File: arch/z80/src/ez80/Toolchain.defs [hunk 5]
```diff
@@ -66,9 +66,19 @@ ARCHASMWARNINGS = -W
...
 
+ifneq ($(CONFIG_DEBUG_NOOPT),y)
+  ARCHOPTIMIZATION += -fno-strict-aliasing
+endif
+
+ifeq ($(CONFIG_FRAME_POINTER),y)
+  ARCHOPTIMIZATION += -fno-omit-frame-pointer -fno-optimize-sibling-calls
+else
+  ARCHOPTIMIZATION += -fomit-frame-pointer
+endif
```
Comment (pkarashchenko) [inline arch/z80/src/ez80/Toolchain.defs:77]: Please move this before the comment above

File: arch/z80/src/ez80/Toolchain.defs [hunk 6]
```diff
@@ -66,9 +66,19 @@ ARCHASMWARNINGS = -W
 
 # Optimization of unused sections
 
+ifneq ($(CONFIG_DEBUG_NOOPT),y)
```
Comment (pkarashchenko) [inline arch/z80/src/ez80/Toolchain.defs:69]: there is `ifeq ($(CONFIG_DEBUG_NOOPT),y)` at line 45. Please merge with that selection.
Or maybe you can add `CONFIG_DEBUG_CUSTOMOPT` and make it the same way as in other files

File: arch/sparc/src/sparc_v8/Toolchain.defs [hunk 7]
```diff
@@ -56,19 +56,29 @@ else
...
+ifeq ($(CONFIG_DEBUG_CUSTOMOPT),y)
+  ARCHOPTIMIZATION += $(CONFIG_DEBUG_OPTLEVEL)
+else ifeq ($(CONFIG_DEBUG_FULLOPT),y)
+  ARCHOPTIMIZATION += -O2
+endif
+
 ifeq ($(CONFIG_DEBUG_CUSTOMOPT),y)
-  MAXOPTIMIZATION := $(CONFIG_DEBUG_OPTLEVEL)
+  ARCHOPTIMIZATION += $(CONFIG_DEBUG_OPTLEVEL)
 endif
```
Comment (pkarashchenko) [inline arch/sparc/src/sparc_v8/Toolchain.defs:71]: Duplicates above

File: boards/sim/sim/sim/scripts/Make.defs (hunk 4 above)
Comment (xiaoxiang781216) [reply boards/sim/sim/sim/scripts/Make.defs:49]: Done.

File: arch/z80/src/ez80/Toolchain.defs (hunk 6 above)
Comment (xiaoxiang781216) [reply arch/z80/src/ez80/Toolchain.defs]: Done.

File: arch/sparc/src/sparc_v8/Toolchain.defs (hunk 7 above)
Comment (xiaoxiang781216) [reply arch/sparc/src/sparc_v8/Toolchain.defs]: Done.

**Comment by xiaoxiang781216** [comment]
ping @pkarashchenko

**Comment by pkarashchenko** [comment]
> ping @pkarashchenko 

PR had merge conflict, so I was not able to merge yesterday. I will merge as soon as CI finishes

**Comment by zouboan** [comment]
@xiaoxiang781216 This is a broken commit, my custom boards can't working properly after i merged with master branch. I'm revert by dichotomy, and finally positioned on this commit.
i delete the following code
ifneq ($(CONFIG_DEBUG_NOOPT),y)
  ARCHOPTIMIZATION += $(MAXOPTIMIZATION) -fno-strict-aliasing
endif

in my custom boards Make.defs, but the problem remains

**Comment by pkarashchenko** [comment]
@zouboan So the code building is fine and the behavior of the board is changed? Or it is a build issue that you are reporting?
If it is a build issue then maybe you can share some more details (base arch and base board that you used initially, Make.defs and maybe some parts of the Makefile), so we can identify the root cause.

**Comment by zouboan** [comment]
@pkarashchenko the behavior of the board is changed, It seems that the board to run very slowly, It seems that the optimization doesn't work. my custom boards config as follow:
#
# Debug Options
#
CONFIG_NDEBUG=y
# CONFIG_DEBUG_ALERT is not set
# CONFIG_DEBUG_FEATURES is not set
# CONFIG_DEBUG_TCBINFO is not set
CONFIG_ARCH_HAVE_STACKCHECK=y
CONFIG_STACK_COLORATION=y
# CONFIG_STACK_CANARIES is not set
CONFIG_ARCH_HAVE_HEAPCHECK=y
# CONFIG_HEAP_COLORATION is not set
CONFIG_DEBUG_SYMBOLS=y
# CONFIG_FRAME_POINTER is not set
CONFIG_ARCH_HAVE_CUSTOMOPT=y
# CONFIG_DEBUG_NOOPT is not set
# CONFIG_DEBUG_CUSTOMOPT is not set
CONFIG_DEBUG_FULLOPT=y
CONFIG_DEBUG_OPT_UNUSED_SECTIONS=y
CONFIG_DEBUG_LINK_MAP=y

**Comment by xiaoxiang781216** [comment]
@zouboan could you run "make V=1" and compare the command line difference between the bad output and the good one?

**Comment by zouboan** [comment]
@xiaoxiang781216 
before this commit:

CC:  irq/irq_attach.c
arm-none-eabi-gcc -c -fno-common -fno-builtin -ffunction-sections -fdata-sections -funwind-tables -Wall -Wstrict-prototypes -Wshadow -Wundef -g -Os -fomit-frame-pointer -ffunction-sections -fdata-sections -g -fno-strict-aliasing -march=armv7e-m -mtune=cortex-m4 -mthumb -mfpu=fpv4-sp-d16 -mfloat-abi=hard -isystem "/home/zouboan/G/NuttX/nuttx/include" -D__NuttX__ -DNDEBUG -D__KERNEL__  -pipe -I "/home/zouboan/G/NuttX/nuttx/sched"   irq/irq_attach.c -o  irq_attach.o

after this commit:

CC:  group/group_find.c
arm-none-eabi-gcc -c -fno-common -fno-builtin -ffunction-sections -fdata-sections -funwind-tables -Wall -Wstrict-prototypes -Wshadow -Wundef -g  -fno-strict-aliasing -march=armv7e-m -mtune=cortex-m4 -mthumb -mfpu=fpv4-sp-d16 -mfloat-abi=hard -isystem "/home/zouboan/G/NuttX/nuttx/include" -D__NuttX__ -DNDEBUG -D__KERNEL__  -pipe -I "/home/zouboan/G/NuttX/nuttx/sched"   group/group_find.c -o  group_find.o

**Comment by xiaoxiang781216** [comment]
The setting come from:
https://github.com/apache/incubator-nuttx/blob/master/arch/arm/src/armv7-m/Toolchain.defs#L60-L68
Please ensure you use `ARCHOPTIMIZATION +=` instead of `ARCHOPTIMIZATION =` in your Make.defs.

**Comment by zouboan** [comment]
@xiaoxiang781216 Thanks for your help, there's a ARCHOPTIMIZATION = -g in my Make.defs , and the problem has been solved now.

**Pull request merged**